### PR TITLE
feat: Add github-upload-public-key module

### PR DESCRIPTION
**Pull request by f0ssel**
Automates uploading Coder public key to Github so users don't have to.

First build:
Screenshot 2024-05-02 at 2 35 35 PM

Subsequent builds:
Screenshot 2024-05-02 at 2 36 02 PM

Github Auth not linked:
Screenshot 2024-05-02 at 2 36 54 PM

Closes https://github.com/coder/customers/issues/542

### CODE REVIEW

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:22]: Even better, you can use `coder external-auth access-token github` to automatically fetch the token for a provider! Then it'd refresh as well.

Comment (Emyrk) [reply github-upload-public-key/run.sh:22]: Yup, `GITHUB_TOKEN` env var method is only because the `gh` cli does not support a more dynamic approach. I've been pushing everyone in the community to not use when they do not have to.

Comment (kylecarbs) [reply github-upload-public-key/run.sh:22]: Yeah, we should remove this from dogfood.

Comment (kylecarbs) [reply github-upload-public-key/run.sh:22]: https://github.com/coder/coder/pull/13106

File: github-upload-public-key/README.md
Comment (matifali) [inline github-upload-public-key/README.md:5]:
```suggestion
maintainer_github: coder
```
We can change it to as official coder module too.

**Comment by f0ssel** [comment]
Working on adding some tests

File: github-upload-public-key/run.sh
Comment (matifali) [inline github-upload-public-key/run.sh:69]: I suggest we generalize to accept custom GitHub URLs for GitHub enterprises customers. 
We can do that by accepting an additional input and default to github.com as fallback. 

Bonus if we can auto detect the GitHub URL from the the configured external auth. 

cc: @Emyrk

File: github-upload-public-key/run.sh
Comment (matifali) [inline github-upload-public-key/run.sh:96]: Same observation here to use a custom GitHub URL.

**Review by matifali — COMMENTED** [review]
I suggest we generalize GitHub URL too by making it an input to the module or better deducing it from the configured GitHub external authentication.

**Review by kylecarbs — APPROVED** [review]
This is absolutely sick!

File: github-upload-public-key/main.tf
Comment (kylecarbs) [inline github-upload-public-key/main.tf:40]: You should be able to pass these as env vars when running the script instead, that way we don't have to have these extra vars.

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:11]: For all of these I might put the env var it's looking for instead. Might reduce some confusion.

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:33]: The output in this scenario will actually be the URL.

See: https://coder.com/docs/v2/latest/cli/external-auth_access-token#description

It'd be nice to output similarly to slackme in this scenario:

https://github.com/coder/modules/blob/main/slackme/slackme.sh#L64

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:39]: I don't think this can happen...

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:40]: This might be helpful for debugging purposes, but I don't think this adds much for the user because of the messages below.

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:42]:
```suggestion
echo "Fetching public key from Coder..."
```

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:60]: Is there a situation where it's not found? I might omit this line.

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:67]:
```suggestion
echo "Fetching public keys from GitHub..."
```

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:88]:
```suggestion
  echo "Your Coder public key is already on GitHub!"
```

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:93]:
```suggestion
echo "Your Coder public key is not in GitHub. Adding it now..."
```

File: github-upload-public-key/run.sh
Comment (kylecarbs) [inline github-upload-public-key/run.sh:114]:
```suggestion
echo "Your Coder public key has been added to GitHub!"
```

**Review by kylecarbs — APPROVED** [review]
Minor language things that I think could be helpful!

File: github-upload-public-key/main.tf
Comment (matifali) [reply github-upload-public-key/main.tf:40]: Yeah I can't think of. Use case where thet will be different from the provider provided values.

Comment (f0ssel) [reply github-upload-public-key/main.tf]: I was able to do this 👍

**Review by matifali — APPROVED** [review]

File: CONTRIBUTING.md
Comment (f0ssel) [reply CONTRIBUTING.md:13]: good addition i meant to make, ty

**Comment by f0ssel** [comment]
I realize there is a conflict between making tests pass and the script actually working in workspaces, I think I need to handle the terraform injection (I didn't know this was actually happening at first) as well as env vars for tests.

**Comment by f0ssel** [comment]
Okay input variable issues are fixed, screenshots are updated, I think this is g2g but I'll wait until the eod to merge so yall can take a look at the last couple changes since review.

**Pull request merged**
